File: internal/domain/entity/phone_challenge.go
package entity

import (
	"github.com/google/uuid"
	"time"
)

type PhoneChallenge struct {
	Id               uuid.UUID
	CreatedAt        time.Time
	PhoneNumber      string
	VerificationCode string
	ExpiredAt        time.Time
}

func NewPhoneChallenge(id uuid.UUID, phoneNumber, verificationCode string, duration time.Duration) (*PhoneChallenge, error) {
	timeNow := time.Now()

	return &PhoneChallenge{
		Id:               id,
		CreatedAt:        timeNow,
		PhoneNumber:      phoneNumber,
		VerificationCode: verificationCode,
		ExpiredAt:        timeNow.Add(duration),
	}, nil
}

File: Domain/Entities/PhoneChallenge.cs
using Domain.Primitives;
using Domain.Shared;

namespace Domain.Entities;

public class PhoneChallenge : AggregateRoot
{
    public Guid Id { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public string PhoneNumber { get; private set; }
    public string VerificationCode { get; private set; }
    public DateTime ExpiredAt { get; private set; }

    private PhoneChallenge(
        Guid id,
        string phoneNumber,
        string verificationCode,
        DateTime createdAt,
        DateTime expiredAt)
    {
        Id = id;
        CreatedAt = createdAt;
        PhoneNumber = phoneNumber;
        VerificationCode = verificationCode;
        ExpiredAt = expiredAt;
    }

    public static Result<PhoneChallenge> Create(
        Guid id,
        string phoneNumber,
        string verificationCode,
        TimeSpan duration)
    {
        var createdAt = DateTime.UtcNow;
        var expiredAt = createdAt.Add(duration);
        return new PhoneChallenge(id, phoneNumber, verificationCode, createdAt, expiredAt);
    }

    public bool IsExpired() => DateTime.UtcNow > ExpiredAt;
}